engine/orders.go

```diff
@@ -204,18 +313,14 @@ func (o *orderManager) Submit(exchName string, newOrder *order.Submit) (*orderSu
 		}
 	}
 
-	exch := GetExchangeByName(exchName)
+	exch := GetExchangeByName(newOrder.Exchange)
 	if exch == nil {
 		return nil, errors.New("unable to get exchange by name")
 	}
-
-	id, err := uuid.NewV4()
-	if err != nil {
-		log.Warnf(log.OrderMgr,
-			"Order manager: Unable to generate UUID. Err: %s\n",
-			err)
-	}
-
+	// Lock at this point to ensure wrapper + websocket order submissions return
+	// before any routines.go order submission competes and this returns an error
+	o.orderStore.m.Lock()
+	defer o.orderStore.m.Unlock()
```

Comment:
So this was added to address a data race where a wrapper websocket request and the datahandler would want to add an order to the orderstore at the same time. There was potential for the wrapper caller to get an "order already exists" error because of that. To address this, I've added the following function:
```
func (w *WebsocketConnection) IsIDWaitingForResponse(id int64) bool {
	w.Lock()
	for k := range w.IDResponses {
		if k == id && w.IDResponses[k] == nil {
			w.Unlock()
			return true
		}
	}
	w.Unlock()
	return false
}
```
Now, when we are doing a wrapper websocket request, we add an ID to the `IDResponses` map with a nil value which says we are awaiting a response.
When a response (or a timeout) occurs, that id will be removed from the map. So now there are no locks used here because we now know when to pass data through the data handler and when not to.

**tl;dr**, removed those locks

Reply:
Awesome work mate!